### PR TITLE
Added 'show' and 'hide' methods to dropdown

### PR DESCRIPTION
**Pull request by galczo5**
Fixes: #22835

### CODE REVIEW

**Review by Johann-S — CHANGES_REQUESTED** [review]
Hi @galczo5 ,

Thank you for your PR 👍 

But can you add a bit of context, why should we need those methods ? What's your use case ?

And can you add unit test for those two methods ?

**Comment by galczo5** [comment]
Hi,
In company we had an issue that we wanted to close all opened dropdowns when user select an option from one of them. We couldn't use `toggle` for it because it would open closed dropdowns. We had to keep state of each dropdown and conditionally call `toggle` method. 
So I've implemented missing `hide` method, and `show` method for consistency.

Later today I'll try to add missing unit tests.

**Comment by galczo5** [comment]
Unit tests pushed. Added missing trigger of `shown` and `hidden` events.

File: js/src/dropdown.js [hunk 1]
```diff
@@ -194,6 +194,61 @@ class Dropdown {
...
+    if (isActive) {
+      return
+    }
+
+    const relatedTarget = {
+      relatedTarget: this._element
+    }
+    const showEvent = $.Event(Event.SHOW, relatedTarget)
+
+    const parent   = Dropdown._getParentFromElement(this._element)
```
Comment (Johann-S) [inline js/src/dropdown.js]: trailing spaces

File: js/src/dropdown.js [hunk 2]
```diff
@@ -194,6 +194,61 @@ class Dropdown {
...
+    if (!isActive) {
+      return
+    }
+
+    const relatedTarget = {
+      relatedTarget: this._element
+    }
+    const hideEvent = $.Event(Event.HIDE, relatedTarget)
+
+    const parent   = Dropdown._getParentFromElement(this._element)
```
Comment (Johann-S) [inline js/src/dropdown.js]: trailing spaces

**Review by Johann-S — CHANGES_REQUESTED** [review]
Can you add more tests please, there are some lines which need tests:

image

- When it's disabled
- When it's active

And there are some line on the `hide` method too

**Comment by galczo5** [comment]
I've just added requested tests. I'm not sure if `preventDefault` tests are written correctly because I'm not very experienced with karma tests. 

It tests are not good, can you give me short description how should it works?
Thanks for your help!

**Comment by Johann-S** [comment]
Thank you that's better, there are just two use cases missings:

- You should check that we do nothing if the dropdown is disabled on show and hide

and it'll be good 👍 

/CC @XhmikosR what do you think about that feature ? Seems interesting to me

**Comment by galczo5** [comment]
There are four tests for disabled dropdowns. One when dropdown is disabled via attribute, second when dropdown is disabled via class. For both methods.

**Comment by galczo5** [comment]
Are we ready for merge?

**Comment by Johann-S** [comment]
I'm waiting for other team members feedbacks

File: js/src/dropdown.js [hunk 3]
```diff
@@ -194,6 +194,61 @@ class Dropdown {
       .trigger($.Event(Event.SHOWN, relatedTarget))
   }
 
+  show() {
+    if (this._element.disabled || $(this._element).hasClass(ClassName.DISABLED)) {
+      return
+    }
+    const isActive = $(this._menu).hasClass(ClassName.SHOW)
```
Comment (XhmikosR) [inline js/src/dropdown.js]: Maybe skip `isActive` here and later one since they are used only once.

**Comment by galczo5** [comment]
@XhmikosR pushed

**Review by Johann-S — CHANGES_REQUESTED** [review]
sorry @galczo5 but I forgot one thing ! 😄 

Can you add those methods in our docs ? (`dropdown.md`)

**Comment by galczo5** [comment]
@Johann-S No problem :) I will edit it today

**Review by Johann-S — APPROVED** [review]

**Comment by galczo5** [comment]
Done :)

**Pull request merged**